aziot/src/internal/check/checks/host_connect_iothub.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use anyhow::{Context, Result};
+use serde::Serialize;
+
+use crate::internal::check::{CheckResult, Checker, CheckerCache, CheckerMeta, CheckerShared};
+
+pub fn host_connect_iothub_checks() -> impl Iterator<Item = Box<dyn Checker>> {
+    let mut v: Vec<Box<dyn Checker>> = Vec::new();
+
+    v.push(Box::new(HostConnectIotHub::new(
+        "host-connect-iothub-amqp",
+        "host can connect to and perform TLS handshake with iothub AMQP port",
+        5671,
+    )));
+    v.push(Box::new(HostConnectIotHub::new(
+        "host-connect-iothub-https",
+        "host can connect to and perform TLS handshake with iothub HTTPS / WebSockets port",
+        443,
+    )));
+    v.push(Box::new(HostConnectIotHub::new(
+        "host-connect-iothub-mqtt",
+        "host can connect to and perform TLS handshake with iothub MQTT port",
+        8883,
+    )));
+
+    v.into_iter()
+}
+
+#[derive(Serialize)]
+pub struct HostConnectIotHub {
+    port_number: u16,
+    iothub_hostname: Option<String>,
+    proxy: Option<String>,
+
+    #[serde(skip)]
+    meta: CheckerMeta,
+}
+
+impl HostConnectIotHub {
+    fn new(id: &'static str, description: &'static str, port_number: u16) -> HostConnectIotHub {
+        HostConnectIotHub {
+            port_number,
+            iothub_hostname: None,
+            proxy: None,
+            meta: CheckerMeta { id, description },
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Checker for HostConnectIotHub {
+    fn meta(&self) -> CheckerMeta {
+        self.meta
+    }
+
+    async fn execute(&mut self, shared: &CheckerShared, cache: &mut CheckerCache) -> CheckResult {
+        self.inner_execute(shared, cache)
+            .await
+            .unwrap_or_else(CheckResult::Failed)
+    }
+}
+
+impl HostConnectIotHub {
+    async fn inner_execute(
+        &mut self,
+        _shared: &CheckerShared,
+        cache: &mut CheckerCache,
+    ) -> Result<CheckResult> {
+        use aziot_identityd_config::ProvisioningType;
+
+        let iothub_hostname = match &unwrap_or_skip!(&cache.cfg.identityd)
+            .provisioning
+            .provisioning
+        {
+            ProvisioningType::Manual {
+                iothub_hostname, ..
+            } => iothub_hostname,
+            _ => return Ok(CheckResult::Ignored),
```

Comment:
The iotedge version of this check gets the hub name even for DPS provisioning if the device has been provisioned previously. Can you do that here? (Either read the backup file like the iotedge version does, or call out to IS, or perhaps Rajeev has a better idea.)

Reply:
Calling out to IS isn't an option, since this check should work regardless of whether or not the IS is running.

What's this backup file you're referring to?

Also, pinging @massand for input

Reply:
>What's this backup file you're referring to?

https://github.com/Azure/iot-identity-service/blob/7e6c062988f26299fc4695d0df55a49e48b839cf/identity/aziot-identityd/src/lib.rs#L144-L145

Reply:
Oh, interesting...

> The iotedge version of this check gets the hub name even for DPS provisioning if the device has been provisioned previously

Wait a sec, unless I'm looking at the wrong code, _does_ it actually do that?
I'm looking at https://github.com/Azure/iotedge/blob/iiot/edgelet/iotedge/src/check/checks/host_connect_dps_endpoint.rs and/or https://github.com/Azure/iotedge/blob/iiot/edgelet/iotedge/src/check/checks/host_connect_upstream.rs

Reply:
Those read `check.iothub_hostname`, not set it. It's set in `well_formed_connection_string.rs`, and the relevant code is https://github.com/Azure/iotedge/blob/iiot/edgelet/iotedge/src/check/checks/well_formed_connection_string.rs#L51-L66

Reply:
Ahhh, I see now. 

Alright, so you're suggesting reading + parsing the `prev_state` backup if `cache.cfg.identityd` isn't available? Yeah, I could do that.

Reply:
>Alright, so you're suggesting reading + parsing the `prev_state` backup if `cache.cfg.identityd` isn't available? Yeah, I could do that.

No, they are different things. The procedure to use them maps one-to-one with the iotedged config and provisioning backup as in the code I linked.

If IS config doesn't exist at all, there's nothing you can do. Give up and move on to the other checks.

If IS config exists, and says it's manual provisioning, you know what the hub name is. The IS config will exist even if the provisioning backup doesn't exist, so for manual provisioning you only need to look at this one.

If IS config exists, and says it's DPS provisioning, the config doesn't tell you the hub name. In this case the only way to get it automatically is to read the provisioning backup. If the user didn't already override the hostname via `--iothub-hostname`, you fail.

Reply:
Okay, I pushed up a partial fix. The only thing that's missing is that additional logic regarding the `--iothub-hostname` flag, which doesn't exist in `aziot`. Let me know if I should implement it, or if it's fine to leave it out.

I'm signing off for the day, but I'll get on it tomorrow.

Reply:
That's true. For `iotedge check` the flag came first, which is why we didn't remove it when we added the fallback to the backup file. But nobody remembered to set the flag anyway, even when the message telling them to set it was staring them in the face. So I'd be fine with not carrying it forward into 1.2

Check with Micah and Venkat, though.

Reply:
Alright, I've re-added the flag. I'll need to update the upstream PR to plumb through the parameter, but it shouldn't be too bad.